File: src_py/services/client_service.py
import uuid
from typing import Dict, Any, List, Optional
from src_py.config import settings
from src_py.db import models
from src_py.crypto.tokens import (
    generate_client_id,
    generate_client_secret,
    hash_secret,
    verify_secret,
    sign_static_token
)


def register_client(
    name: str,
    audience: str,
    allowed_redirect_uris: Optional[List[str]] = None,
    client_type: str = "confidential",
    ip_address: Optional[str] = None
) -> Dict[str, Any]:
    """Registers a new MCP client via Admin UI or RFC 7591 Dynamic Client Registration."""
    if not name or not audience:
        raise ValueError("Both name and audience are required")

    clean_audience = audience.strip()
    existing = models.get_client_by_audience(clean_audience)
    if existing:
        raise ValueError(f'An MCP client with audience "{clean_audience}" already exists')

    client_id = generate_client_id(name)
    raw_secret = generate_client_secret()
    secret_hash = hash_secret(raw_secret)
    client_uuid = str(uuid.uuid4())
    redirect_uris = allowed_redirect_uris or []

    client = models.create_client(
        id=client_uuid,
        name=name.strip(),
        client_id=client_id,
        client_secret_hash=secret_hash,
        audience=clean_audience,
        allowed_redirect_uris=redirect_uris,
        client_type=client_type
    )

    models.add_token_event(
        event_type="registered",
        client_id=client_id,
        audience=clean_audience,
        mode="admin_or_dcr",
        details={"name": client["name"], "clientType": client_type},
        ip_address=ip_address
    )

    return {
        "client": client,
        "clientSecret": raw_secret  # Revealed ONCE to caller
    }


def generate_static_token_for_client(
    client_id: str,
    days: Optional[int] = None,
    ip_address: Optional[str] = None
) -> Dict[str, Any]:
    """Generates a Mode 2 static long-lived token for an MCP client."""
    client = models.get_client_by_client_id(client_id)
    if not client:
        raise ValueError("Client not found")
    if client.get("revoked"):
        raise ValueError("Cannot issue token for a revoked client")

    token_data = sign_static_token(
        client_id=client["client_id"],
        audience=client["audience"],
        days=days
    )

    models.add_token_event(
        event_type="issued",
        client_id=client["client_id"],
        audience=client["audience"],
        mode="static_token",
        details={"jti": token_data["jti"], "days": days, "expiresIn": token_data["expires_in"]},
        ip_address=ip_address
    )

    return {
        "token": token_data["token"],
        "jti": token_data["jti"],
        "expiresIn": token_data["expires_in"],
        "audience": client["audience"],
        "clientId": client["client_id"],
        "tokenType": "Bearer"
    }


def authenticate_client(client_id: str, client_secret: str) -> Optional[Dict[str, Any]]:
    """Authenticates client credentials."""
    client = models.get_client_by_client_id(client_id)
    if not client or client.get("revoked"):
        return None

    if not verify_secret(client_secret, client.get("client_secret_hash", "")):
        return None

    return client


def revoke_client(client_id_or_id: str, ip_address: Optional[str] = None) -> Dict[str, Any]:
    """Revokes an MCP client immediately."""
    client = models.get_client_by_id(client_id_or_id) or models.get_client_by_client_id(client_id_or_id)
    if not client:
        raise ValueError("Client not found")

    updated = models.update_client_revocation(client["id"], True)

    models.add_token_event(
        event_type="revoked",
        client_id=client["client_id"],
        audience=client["audience"],
        mode="admin",
        details={"action": "revoke_client"},
        ip_address=ip_address
    )

    return updated


def unrevoke_client(client_id_or_id: str, ip_address: Optional[str] = None) -> Dict[str, Any]:
    """Restores/unrevokes an MCP client."""
    client = models.get_client_by_id(client_id_or_id) or models.get_client_by_client_id(client_id_or_id)
    if not client:
        raise ValueError("Client not found")

    updated = models.update_client_revocation(client["id"], False)

    models.add_token_event(
        event_type="restored",
        client_id=client["client_id"],
        audience=client["audience"],
        mode="admin",
        details={"action": "unrevoke_client"},
        ip_address=ip_address
    )

    return updated

File: src/services/clientService.js
const crypto = require('crypto');
const models = require('../db/models');
const {
  generateClientId,
  generateClientSecret,
  hashSecret,
  verifySecret,
  signStaticToken
} = require('../crypto/tokens');
const config = require('../config');

const clientService = {
  /**
   * Registers a new MCP client.
   * Can be invoked via Admin UI or RFC 7591 Dynamic Client Registration.
   */
  async registerClient({
    name,
    audience,
    allowedRedirectUris = [],
    clientType = 'confidential',
    ipAddress = null
  }) {
    if (!name || !audience) {
      throw new Error('Both name and audience are required');
    }

    // Ensure audience is clean and unique
    const cleanAudience = audience.trim();
    const existingAud = models.getClientByAudience(cleanAudience);
    if (existingAud) {
      throw new Error(`An MCP client with audience "${cleanAudience}" already exists`);
    }

    const clientId = generateClientId(name);
    const rawSecret = generateClientSecret();
    const secretHash = await hashSecret(rawSecret);
    const id = crypto.randomUUID();

    const client = models.createClient({
      id,
      name: name.trim(),
      clientId,
      clientSecretHash: secretHash,
      audience: cleanAudience,
      allowedRedirectUris,
      clientType
    });

    models.addTokenEvent({
      clientId,
      audience: cleanAudience,
      eventType: 'registered',
      mode: 'admin_or_dcr',
      details: { name: client.name, clientType },
      ipAddress
    });

    return {
      client,
      clientSecret: rawSecret // Revealed ONCE to the caller
    };
  },

  /**
   * Generates a Mode 2 static long-lived token for an MCP client.
   */
  generateStaticToken(clientId, days = config.staticTokenExpiryDays, ipAddress = null) {
    const client = models.getClientByClientId(clientId);
    if (!client) {
      throw new Error('Client not found');
    }
    if (client.revoked) {
      throw new Error('Cannot issue token for a revoked client');
    }

    const { token, jti, expiresIn } = signStaticToken({
      clientId: client.client_id,
      audience: client.audience,
      days
    });

    models.addTokenEvent({
      clientId: client.client_id,
      audience: client.audience,
      eventType: 'issued',
      mode: 'static_token',
      details: { jti, days, expiresIn },
      ipAddress
    });

    return {
      token,
      jti,
      expiresIn,
      audience: client.audience,
      clientId: client.client_id,
      tokenType: 'Bearer'
    };
  },

  /**
   * Authenticates client credentials.
   */
  async authenticateClient(clientId, clientSecret) {
    const client = models.getClientByClientId(clientId);
    if (!client) return null;
    if (client.revoked) return null;

    const valid = await verifySecret(clientSecret, client.client_secret_hash);
    if (!valid) return null;

    return client;
  },

  /**
   * Revokes an MCP client immediately.
   */
  revokeClient(id, ipAddress = null) {
    const client = models.getClientById(id);
    if (!client) throw new Error('Client not found');

    models.updateClientRevocation(id, true);

    models.addTokenEvent({
      clientId: client.client_id,
      audience: client.audience,
      eventType: 'revoked',
      mode: 'admin',
      details: { action: 'revoke_client' },
      ipAddress
    });

    return models.getClientById(id);
  },

  /**
   * Restores/unrevokes an MCP client.
   */
  unrevokeClient(id, ipAddress = null) {
    const client = models.getClientById(id);
    if (!client) throw new Error('Client not found');

    models.updateClientRevocation(id, false);

    models.addTokenEvent({
      clientId: client.client_id,
      audience: client.audience,
      eventType: 'restored',
      mode: 'admin',
      details: { action: 'unrevoke_client' },
      ipAddress
    });

    return models.getClientById(id);
  }
};

module.exports = clientService;
